deepchem/models/torch_models/gnn.py

```diff
@@ -0,0 +1,360 @@
+            else:
+                pool = Set2Set(self.emb_dim, processing_steps=set2setiter)
+
+        if self.graph_pooling == "set2set":
+            mult = 2
+        else:
+            mult = 1
+
+        if self.JK == "concat":
+            head = torch.nn.Linear(mult * (self.num_layer + 1) * self.emb_dim,
+                                   self.num_tasks)
+        else:
+            head = torch.nn.Linear(mult * self.emb_dim, self.num_tasks)
+
+        components = {
+            'atom_type_embedding':
+                torch.nn.Embedding(num_atom_type, self.emb_dim),
+            'chirality_embedding':
+                torch.nn.Embedding(num_chirality_tag, self.emb_dim),
+            'gconvs':
+                encoders,
+            'batch_norms':
+                batch_norms,
+            'pool':
+                pool,
+            'head':
+                head
+        }
+        self.gnn = GNN(components['atom_type_embedding'],
+                       components['chirality_embedding'], components['gconvs'],
+                       components['batch_norms'], self.dropout, self.JK)
+        self.gnn_head = GNN_head(components['pool'], components['head'])
+        return components
+
+    def build_model(self):
+        if self.task == "edge_pred":  # unsupervised task, does not need pred head
+            return self.gnn
+        else:
+            return torch.nn.Sequential(self.gnn, self.gnn_head)
+
+    def loss_func(self, inputs, labels, weights):
+        if self.task == "edge_pred":
+            return self.edge_pred_loss(inputs, labels, weights)
+
+    def edge_pred_loss(self, inputs, labels, weights):
+        node_emb, _ = self.model(
+            inputs)  # node_emb shape == [num_nodes x emb_dim]
+
+        positive_score = torch.sum(node_emb[inputs.edge_index[0, ::2]] *
+                                   node_emb[inputs.edge_index[1, ::2]],
+                                   dim=1)
+        negative_score = torch.sum(node_emb[inputs.negative_edge_index[0]] *
+                                   node_emb[inputs.negative_edge_index[1]],
+                                   dim=1)
+
+        loss = self.criterion(
+            positive_score, torch.ones_like(positive_score)) + self.criterion(
+                negative_score, torch.zeros_like(negative_score))
+        return (loss * weights[0]).mean()
+
+    def _prepare_batch(self, batch):
+        """
+        Prepares the batch for the model by converting the GraphData numpy arrays to torch tensors and moving them to the device.
+        """
+        inputs, labels, weights = batch
+        inputs = BatchGraphData(inputs[0]).numpy_to_torch(self.device)
+        inputs = self.neg_edges(inputs)
+
+        _, labels, weights = super()._prepare_batch(([], labels, weights))
+
+        if (len(labels) != 0) and (len(weights) != 0):
+            labels = labels[0]
+            weights = weights[0]
+
+        return inputs, labels, weights
+
+
+class NegativeEdge:
```

Comment:
Following up on our offline conversation, this should probably be in utils as we discussed

Reply:
modified as a helper function